File: solutions/go/scrabble-score/scrabble_score.go
package scrabble

import "strings"

// Score returns the Scrabble score for a given word.
func Score(word string) (sum int) {
	word = strings.ToUpper(word)
	for _, c := range word {
		sum += scoreLetter(c)
	}
	return
}

func scoreLetter(char rune) int {
	switch char {
	case 'A', 'E', 'I', 'O', 'U', 'L', 'N', 'R', 'S', 'T':
		return 1
	case 'D', 'G':
		return 2
	case 'B', 'C', 'M', 'P':
		return 3
	case 'F', 'H', 'V', 'W', 'Y':
		return 4
	case 'K':
		return 5
	case 'J', 'X':
		return 8
	case 'Q', 'Z':
		return 10
	default:
		return 0
	}
}

File: solutions/r/scrabble-score/scrabble-score.R
# Returns the Scrabble score for a given word.
scrabble_score <- function(input) {
    if (input == "") {
        return(0)
    }

    score_letter <- Vectorize(function(char) {
        if (char %in% c("A", "E", "I", "O", "U", "L", "N", "R", "S", "T")) {
            1
        } else if (char %in% c("D", "G")) {
            2
        } else if (char %in% c("B", "C", "M", "P")) {
            3
        } else if (char %in% c("F", "H", "V", "W", "Y")) {
            4
        } else if (char %in% c("K")) {
            5
        } else if (char %in% c("J", "X")) {
            8
        } else if (char %in% c("Q", "Z")) {
            10
        } else {
            0
        }
    })

    (input |> toupper()
        |> strsplit("") |> unlist()
        |> score_letter() |> sum())
}
